### DISCUSSION TITLE
Can't get basic authentication working or other issue

### DISCUSSION THREAD

**Discussion by mdrich-com** [btschwertfeger/python-kraken-sdk#226] [Q&A]
I feel like an idiot. I copied the demo code for API V2 here:
https://python-kraken-sdk.readthedocs.io/en/stable/src/examples/rest_ws_samples/spot_ws_examples.html

and here in a separate file:
https://python-kraken-sdk.readthedocs.io/en/stable/src/examples/trading_bot_templates/spot_bot_templates.html

I am having issues getting things to work in each of those examples. In the `spot_bot_templates.html `I can't get past

`cannot import name 'KrakenAuthenticationError' from 'kraken.exceptions'` 

unless I delete that whole section. Once that's fixed, I get errors on authentication checks. My keys are fine. I just created the key for this today.

On `spot_ws_examples.html` I can get the websocket running and pull out the information I need, present it and do other actions with the inbound data, but I can't get my orders sent using the examples from https://python-kraken-sdk.readthedocs.io/en/stable/src/spot/websockets.html#kraken.spot.KrakenSpotWSClientV2

Attempted "sell" code:

`await client_auth.send_message(

    message={

        "method": "add_order",

        "params": {

            "limit_price": 1234.56,

            "order_type": "limit",

            "order_userref": 123456789,

            "order_qty": 1.0,

            "side": "buy",

            "symbol": "BTC/USD",

        },

    }

)`

I managed to get it to run without an error and it returned "None" as a message... I think. My output was "None," but I don't know if it was the response from Kraken or not. More changes using other example code in the readthedocs.io site left me with a number of errors that evenutally leads me to believe the keys aren't being passed or I'm not authenticated. How do I get authenticated? I have my Public and Private keys in their fields.

**Comment by mdrich-com** [comment]
Update... attempting to ignore the above and do REST I get the following issues.

I realize this might need to be broken up into individual threads, but I'm also not sure if it's a single issue (installation, corrupted install, something I'm missing that is causing multiple issues) that is affecting multiple points or if there are separate issues. I can't imagine it being this many separate issues though because I don't see other issues on simple tasks like this in the QA tags here, and I would expect a lot of threads if there were "a lot of issues," but python-kraken-sdk seems to be made very well... just not running well on my computers.

```
print(trade.create_order(  #Attempt to make trade and receive response - throws error on volume
         ordertype="limit",
         side="sell",
         volume = "1",
         pair="DOGE/USD",
         price=1
    ))
```

Gives the following:

```kraken.exceptions.KrakenException.KrakenInvalidArgumentsError: The request payload is malformed, incorrect or ambiguous.
Details: {'error': ['EGeneral:Invalid arguments:volume']}
```

Taking volume out gives the following:
```
TypeError: Trade.create_order() missing 1 required positional argument: 'volume'
```

**Comment by btschwertfeger** [comment]
Hey @mdrich-com, thanks for reaching out!

which version of the python-kraken-sdk are you using? v2.2.0 is the latest version which I'm using below.

I just tried both template bots you mentioned at https://python-kraken-sdk.readthedocs.io/en/stable/src/examples/trading_bot_templates/spot_bot_templates.html and they just run through. 

You may forgot to provide the environment variables SPOT_API_KEY and SPOT_SECRET_KEY (https://github.com/btschwertfeger/python-kraken-sdk/blob/v2.2.0/examples/spot_trading_bot_template_v2.py#L235-L236)?

```python
SPOT_API_KEY=<our-public-key> SPOT_SECRET_KEY=<your-secret-key> python3 examples/spot_trading_bot_template_v2.py

2024/06/05 05:30:24 spot_trading_bot_template_v2,line: 210     INFO | Client credentials are valid.
2024/06/05 05:30:24 connectors,line: 191     INFO | Websocket start connect/reconnect
2024/06/05 05:30:24 connectors,line: 191     INFO | Websocket start connect/reconnect
2024/06/05 05:30:25 connectors,line: 514     INFO | Recover authenticated subscriptions []: waiting
2024/06/05 05:30:25 connectors,line: 121     INFO | Websocket connected!
2024/06/05 05:30:25 connectors,line: 521     INFO | Recover authenticated subscriptions []: done
2024/06/05 05:30:25 spot_trading_bot_template_v2,line: 76     INFO | {'channel': 'status', 'data': [{'api_version': 'v2', 'connection_id': 9218230092259195900, 'system': 'online', 'version': '2.0.5'}], 'type': 'update'}
2024/06/05 05:30:26 connectors,line: 514     INFO | Recover public subscriptions []: waiting
2024/06/05 05:30:27 connectors,line: 121     INFO | Websocket connected!
2024/06/05 05:30:27 connectors,line: 521     INFO | Recover public subscriptions []: done
2024/06/05 05:30:27 spot_trading_bot_template_v2,line: 76     INFO | {'channel': 'status', 'data': [{'api_version': 'v2', 'connection_id': 16948663002661699511, 'system': 'online', 'version': '2.0.5'}], 'type': 'update'}
``` 

Alternativly you can just hardcode the API keys if passing via the environment is not the desired way.

---

About the `None` return: `send_message` will always return `None` since websockets are like this. You have an open connection on which you receive messages - and can send ones. If you successuflly submitted an order or sent any message via `send_message`, you will never get a response by the function, but receive one by the incomming "steam" of messages. Have a look at the `on_message` function which receives all messages sent by Kraken (https://github.com/btschwertfeger/python-kraken-sdk/blob/master/examples/spot_trading_bot_template_v2.py#L70-L78). You should be able to see either a success of failure notification with further information.

--- 

For your error regarding the volume - of course, the volume is required when placing an order. I'm not 100% sure why you experience this error but It may has to do with the size of your order. Kraken's error messages are not that good and not consistent in any case. One can't open an order with this small volume (1x1DOGE=$1). You can review the minimum order sizes here: https://support.kraken.com/hc/en-us/articles/205893708-Minimum-order-size-volume-for-trading or get them via `from kraken.spot import Market; Market().get_asset_pairs(pair="DOGE/USD")` and then access the `ordermin` entry (https://python-kraken-sdk.readthedocs.io/en/stable/src/spot/rest.html#kraken.spot.Market.get_asset_pairs).

**Reply by mdrich-com** [reply]
Thank you for the answers.

It looks like pip will only install up to 1.6.2 of python-kraken-sdk. That may explain why I have "ImportError: cannot import name 'KrakenAuthenticationError' " until that line is removed


I had not replaced the lines about environment variables entirely in my changes and that was probably it. I also now know why I haven't seen any new messages. I didn't know Websockets didn't offer a dedicated response like REST and I've been filtering my visible messages to only the data I need for the testing. I don't know websockets, but I need them for faster trading now that I've gotten rid of Coinbase.

As for the Doge minimums, I don't know when they changed it to 80, but it's been a while since I've attempted Kraken anything. I've moved my test up above the minimum now. I thought the error was telling me it's a bad criteria altogether as if I could not specify volumes for limit orders, not an incorrect minimum.



For a "Clean slate" I copied the trading bot example again back into a file and added my keys. Everything appears to be working nicely (except the error above that I've worked around for now)

I also modified the order message to the following to test:

```
 await self.send_message(

             message={

                 "method": "add_order",

                 "params": {

                     "limit_price": 2,

                     "order_type": "limit",

                     "order_userref": 123456789,

                     "order_qty": 80.0,

                     "side": "sell",

                     "symbol": "DOGE/USD",

                     "validate": True,

                 },

             }

         )
```

There are no errors, but I now receive this in response to what I assume is my test limit order of now 80 doge at $2 for sale:

2024/06/05 01:15:32 spot trading template,line: 112     INFO | {'method': 'add_order', 'result': {'order_id': '', 'order_userref': 123456789}, 'success': True, 'time_in': '2024-06-05T05:15:32.043596Z', 'time_out': '2024-06-05T05:15:32.044378Z'}

I see no orders on Kraken's interface online. I have a lot of Doge available for sale, and I don't see any errors now. I just don't see the order online or in query with REST. There's also no order_id in the message

Thank you for your help on this. I've been wanting to get my algorithm back up and running since leaving Coinbase, but Kraken has been rough going so far on the API.

As soon as ordering can be tested to work I can add in the real buys/sells hopefully.

**Reply by btschwertfeger** [reply]
Hey @mdrich-com, the limitation to that old pyhton-kraken-sdk may come form the Python version you use. They sdk requires Python 3.11 since version 2.0.0.

--- 
It is not showing any placed order, since you passed `validate=True` which is a "dry-run". So you can use it to check if an order could be placed - but it is actually not placing the order. Just don't pass the validate parameter or pass `validate=False` and you will see your placed order.

**Reply by mdrich-com** [reply]
I'm just dumb I guess lol I see now the "validate" boolean on the Kraken APIV2 page. Had to search for it. I thought it was validate as in "give me a receipt" in-essence like some other sites. Thank you!

I did manage to get Python updated even though it didn't want to. I'll try the complete, with KrakenAuthenticationError imports, later today. 

I need to get more into Python again lol... especially for APIs. Thank you for this help. Now I can get the stuff I actually know in there... algorithms and data crunching (until I inevitably need assistance with something like the CSV library from someone else).

Hopefully I won't have any more issues, and I don't see any more possible now that I am authenticated, able to trade and able to read messages I need so far.

**Comment by adamantmm** [comment]
`KrakenAuthenticationError` import failures usually mean the installed package and the docs are on different major versions. Quick sanity check:

```bash
python -c "import kraken; print(kraken.__version__)"
pip show python-kraken-sdk
```

If you are on v2.x, pull the exception from wherever that version actually exports it (or catch the base API error) instead of copying an older docs snippet.

Also worth isolating REST first with a single signed `get_account_balance` call before wiring the websocket bot template — that tells you whether the problem is keys/permissions vs the WS example itself. API keys need “Query funds” / trade permissions matching what the template does; a key created with withdraw-only or no funding permission fails in confusing places.
